### ISSUE TITLE
Negative value in nodetool listsnapshots

### ISSUE THREAD

**Issue by mdevan** [scylladb/scylla#2104]
Scylla version 1.5.1-20170216.a00c273-ubuntu1 on Ubuntu 16.04.

Nodetool snapshot list shows negative value for "Size on disk" for a certain keyspace. The actual disk size reported by a "du -bc *" in the snapshot directory for that keyspace is 8516162710 bytes. Value reported is -73772455 bytes.

```
$ nodetool listsnapshots
Snapshot Details:
Snapshot name       Keyspace name                Column family name           True size          Size on disk
1487737788617       xxx1                         yyy1                  24.35 KB           186.32 KB
1487737788617       system                       size_estimates               1.01 MB            1.01 MB
1487737788617       system                       compaction_history           0 bytes            68.19 KB
1487737788617       system                       schema_usertypes             0 bytes            10.4 KB
1487737788617       system                       schema_triggers              0 bytes            10.4 KB
1487737788617       system                       IndexInfo                    0 bytes            9.12 KB
1487737788617       system                       schema_columns               0 bytes            71.56 KB
1487737788617       system                       range_xfers                  0 bytes            0 bytes
1487737788617       system                       sstable_activity             0 bytes            0 bytes
1487737788617       system_traces                sessions                     0 bytes            0 bytes
1487737788617       xxx2                         yyy2                   72.59 MB           -73772455 bytes
1487737788617       system_traces                events                       0 bytes            0 bytes
1487737788617       system                       batchlog                     0 bytes            0 bytes
1487737788617       system                       local                        34.47 KB           34.47 KB
1487737788617       system                       paxos                        0 bytes            0 bytes
1487737788617       system                       peer_events                  0 bytes            0 bytes
1487737788617       system                       peers                        39.21 KB           39.21 KB
1487737788617       system_traces                node_slow_log                0 bytes            0 bytes
1487737788617       system                       schema_keyspaces             15.91 KB           15.91 KB
1487737788617       system                       schema_columnfamilies        0 bytes            76.36 KB
1487737788617       system                       compactions_in_progress      0 bytes            0 bytes
1487737788617       system                       hints                        0 bytes            0 bytes

Total TrueDiskSpaceUsed: 4.07 GB

$ nodetool version
ReleaseVersion: 2.1.8

$ sudo dpkg -s scylla-server | grep Version
Version: 1.5.1-20170216.a00c273-ubuntu1
```

```
root@scylla-2:/var/lib/scylla/data/xxx2/yyy2-9d008d50bc3011e682fe000000000000/snapshots/1487737788617# du -bc * | grep total
8516162710	total
```

**Comment by tzach** [comment]
@amnonh ?

**Comment by gleb-cloudius** [comment]
On Wed, Feb 22, 2017 at 01:03:34AM -0800, Mahadevan wrote:
> Scylla version 1.5.1-20170216.a00c273-ubuntu1 on Ubuntu 16.04.
> 
> Nodetool snapshot list shows negative value for "Size on disk" for a certain keyspace. The actual disk size reported by a "du -bc *" in the snapshot directory for that keyspace is 8516162710 bytes. Value reported is -73772455 bytes.
> 
> ```
> $ nodetool listsnapshots
> Snapshot Details:
> Snapshot name       Keyspace name                Column family name           True size          Size on disk
> 1487737788617       xxx1                         yyy1                  24.35 KB           186.32 KB
> 1487737788617       system                       size_estimates               1.01 MB            1.01 MB
> 1487737788617       system                       compaction_history           0 bytes            68.19 KB
> 1487737788617       system                       schema_usertypes             0 bytes            10.4 KB
> 1487737788617       system                       schema_triggers              0 bytes            10.4 KB
> 1487737788617       system                       IndexInfo                    0 bytes            9.12 KB
> 1487737788617       system                       schema_columns               0 bytes            71.56 KB
> 1487737788617       system                       range_xfers                  0 bytes            0 bytes
> 1487737788617       system                       sstable_activity             0 bytes            0 bytes
> 1487737788617       system_traces                sessions                     0 bytes            0 bytes
> 1487737788617       xxx2                         yyy2                   72.59 MB           -73772455 bytes
> 1487737788617       system_traces                events                       0 bytes            0 bytes
> 1487737788617       system                       batchlog                     0 bytes            0 bytes
> 1487737788617       system                       local                        34.47 KB           34.47 KB
> 1487737788617       system                       paxos                        0 bytes            0 bytes
> 1487737788617       system                       peer_events                  0 bytes            0 bytes
> 1487737788617       system                       peers                        39.21 KB           39.21 KB
> 1487737788617       system_traces                node_slow_log                0 bytes            0 bytes
> 1487737788617       system                       schema_keyspaces             15.91 KB           15.91 KB
> 1487737788617       system                       schema_columnfamilies        0 bytes            76.36 KB
> 1487737788617       system                       compactions_in_progress      0 bytes            0 bytes
> 1487737788617       system                       hints                        0 bytes            0 bytes
> 
> Total TrueDiskSpaceUsed: 4.07 GB
> 
> $ nodetool version
> ReleaseVersion: 2.1.8
> 
> $ sudo dpkg -s scylla-server | grep Version
> Version: 1.5.1-20170216.a00c273-ubuntu1
> ```
> 
> ```
> root@scylla-2:/var/lib/scylla/data/xxx2/yyy2-9d008d50bc3011e682fe000000000000/snapshots/1487737788617# du -bc * | grep total
> 8516162710	total
> ```
(gdb) print (signed int)8516162710
$1 = -73771882

--
			Gleb.

**Comment by mdevan** [comment]
The bug is here:

https://github.com/scylladb/scylla-jmx/blob/master/src/main/java/com/scylladb/jmx/api/APIClient.java#L475-L476

```
                SnapshotDetailsTabularData.from(key, obj.getString("ks"), obj.getString("cf"), obj.getInt("total"),
                        obj.getInt("live"), data);
```

The two getInt() calls truncate the number to 32-bit integers.

**Comment by amnonh** [comment]
I already has a patch for it, will send it soon

**Comment by penberg** [comment]
Fixed in master and backported to 1.7, 1.6, and 1.5.
